### ISSUE TITLE
CSS Modules styles defined in wrong order in development mode

### ISSUE THREAD

**Issue by jensmeindertsma** [vercel/next.js#13092]
# Bug report
## Describe the bug
In development mode, when a page imports styles from a `.module.css` file ( `import styles from "./home.module.css"` ) and then imports another component after that ( `import { Container } from "../components/Container"` ), and passes the imported styles to the component like `<Container className={styles.blue}>`, these styles get overriden by the imported component. This component has it's own styles and passes the `className` prop to a DOM element:
```js
import styles from "./Container.module.css";

function classes(...classNames) {
  const joined = classNames.join(' ');
  return joined;
}

export const Container = ({
  children,
  className= "",
}) => {
  return (
    <div className={classes(styles.container, className)}>
	 {children}
    </div>
  );
};
```
The `Container` has the following style:
```css
.container {
	background-color: yellow;
}
```
But the page provides the following style:
```css
.blue {
	background-color: blue;
}
```

In development, the styles are injected into the head as follows:
```html
<style>
.index_blue__3Iu-s {
	background-color: blue;
}
</style>
<style>
.Container_container__1UmhU {
	background-color: yellow;
}
</style>
```
In production, these styles are combined into one stylesheet:
```js
.Container_container__1UmhU {
    background-color: #ff0
}

.index_blue__3Iu-s {
    background-color: #00f
}
```
As you can see, there is a difference in order. This causes the inconsistency across environments with CSS ( yellow in development, blue in production ).

I believe that the production result is the correct one, as you should be able to override the component styles.

I'd be happy to try to contribute to a fix. However, I am a beginner developer so I might need some guidance.
## To Reproduce

Clone my [reproduction repository](https://github.com/jensmeindertsma/next-issue-css) and run `npm run dev` to see the styles applied incorrectly, and run `npm run build; npm run start` to see that the styles work as intended.

## Expected behavior

I would expect that in both development and production mode the classes get defined in the same order so that the styles are applied correctly.

## System information

- OS: Windows 10
- Browser: Chrome  81.0.4044.138
- Version of Next.js: 9.4.1
- Version of Node.js: 12.16.1

## More information
The style tags are injected by the code [in this file](https://github.com/zeit/next.js/blob/canary/packages/next/build/webpack/config/blocks/css/loaders/client.ts).  It uses `style-loader` in development and the `MiniCssExtractPlugin` in production. What is the reason as to why it isn't used in development mode?

<sub>[NEXT-1386](https://linear.app/vercel/issue/NEXT-1386/css-modules-styles-defined-in-wrong-order-in-development-mode)</sub>

**Comment by jensmeindertsma** [comment]
I've now noticed that this was already a problem in 9.3.6. I haven't tested earlier versions. ( I only just started using CSS Modules )

**Comment by jensmeindertsma** [comment]
Further investigation: this can be fixed by moving the `import styles from "./index.module.css" to the bottom of the imports in the page file:
```js
// Before
import styles from "./index.module.css";
import { Container } from "../components/Container";

// After
import { Container } from "../components/Container";
import styles from "./index.module.css";
```
Perhaps Next.js could detect default imports from "<filename>.module.<css,scss>" files and move them to the bottom of all the import statements in the file?

**Comment by jensmeindertsma** [comment]
Also, when you edit the name of the class in `Container.js` and its `.module.css` file from `.container` to `.purple` for example, it applies the styles in a new `<style>` tag that gets placed _below_ the `<style>` with the page styles. The logic that does this would also need some changes to place the new styles above the page styles so that the order stays correct. 

I'd be happy to try to contribute a fix. However, I am a beginner developer so I might need some guidance.

**Comment by jensmeindertsma** [comment]
This can be fixed by moving the import from a CSS file to the end of the import statements in the file. In development, the styles are separated over multiple `<style>` tags. They look like this ( actual order ):
```html
<style>
.index_blue__3Iu-s {
	background-color: blue;
}
</style>
<style>
.Container_container__1UmhU {
	background-color: yellow;
}
</style>
```
In production, these styles are combined into one stylesheet:
```css
.Container_container__1UmhU {
    background-color: #ff0
}

.index_blue__3Iu-s {
    background-color: #00f
}
```
As you can see, there is a difference in order. This causes the inconsistency across environments with CSS.

I'd be happy to try to contribute to a fix. However, I am a beginner developer so I might need some guidance.

**Comment by jensmeindertsma** [comment]
@Timer I rewrote the issue as it wasn't that clear previously.

**Comment by jensmeindertsma** [comment]
Even moving the import to the bottom of the file doesn't work perfectly, because changing the component className still moves that class below the page styles. 

I'd be happy to work on this, but I need some direction. I tried using `MiniCssExtractPlugin` that's already used in production also for development mode but that causes some errors too. Could someone explain why there is a different approach to styles across development and production mode? @Timer

**Comment by schwamic** [comment]
It looks like the classes are called in alphabetical order, which does not correspond to the order of 'className'.

**Comment by petewarman** [comment]
I think the order of the imports matters - in the example page you have 
```
import styles from "./index.module.css";
import { Container } from "../components/Container";
```
So I would expect the source order of the compiled css to be 
1) page styles
2) component styles

Therefore in your example the component style should override the page styles and the background should be yellow. ie. the dev mode is correct. Worryingly it's the production mode that gets it wrong.

**Comment by jensmeindertsma** [comment]
@petewarman Moving the styles import to after the component import works, until I modify the component too much (changing the name of the class it applies to the div). Then the styles of the component are re-inserted after the page styles despite the page styles import being the last import in the file

**Comment by softcraftkmh** [comment]
@petewarman i'm also having the issue too
as for me, i'm using materialize css library 
the library css classes are overriding my class
for example, in className={"btn " + styles.custom_btn}
not all attributes from my custom_btn module scss is not working because they being overridden by global btn class.

**Comment by giuseppeg** [comment]
This is likely an generic issue with CSS Modules and not Next.js specific https://twitter.com/giuseppegurgone/status/1084921816030896135

**Comment by jensmeindertsma** [comment]
You might be right. Although I do still think the loading order should be **consistent** across environments

**Comment by giuseppeg** [comment]
fair point

**Comment by jensmeindertsma** [comment]
Going to keep this open because this is something that Next.js could improve. I'm willing to help work on it.

**Comment by kylemh** [comment]
I'm not even sure what to do as a workaround for this.

I think an easy solution would be to automatically apply page-specific identifiers as part of the css modules hash generation, but the ability to configure that isn't exposed.

In my app today, I guess I could pass the `style` attribute and inline my styles, but I have a heavy reliance on PostCSS stuff that can't be leveraged in my JS.

I tried using `classNames` with dynamic values and `.bind()` to no avail... Even `!important` wont work because if more pages are passing one-offs to the same components, those will be fighting eachother.

**Comment by kylemh** [comment]
For [this deployment](https://operation-code-7g9oz802l.vercel.app/) you can hover on one of these three image buttons, and see the whole page's content shift:

![](https://user-images.githubusercontent.com/9523719/87836930-b9482380-c846-11ea-9c13-af832582e181.png)

Each of those images is a next/link, so a route is pre-fetched... including [it's stylesheet](https://github.com/OperationCode/front-end/blob/main/pages/get_involved.js#L18). The styles on that page end up overriding conditional styles or styles passed in via a className prop, for a component used on both pages [(namely, the "HeroBanner")](https://github.com/operationCode/front-end/blob/main/components/HeroBanner/HeroBanner.js#L27). 

Basically, prefetched stylesheets have the potential to obliterate rules on the current page.

I'm not sure what can be done aside from making a route-named CSS class and wrapping each pages styles in it as part of the build-time step.

**Comment by kylemh** [comment]
Ah shit, @Timer I think my issue and @jensmeindertsma are actually different errors. My problem is happening in production.

**Comment by kylemh** [comment]
@Timer  https://github.com/vercel/next.js/issues/15284

**Comment by Kulakoff1988** [comment]
> I think the order of the imports matters - in the example page you have
> 
> ```
> import styles from "./index.module.css";
> import { Container } from "../components/Container";
> ```
> 
> So I would expect the source order of the compiled css to be
> 
> 1. page styles
> 2. component styles
> 
> Therefore in your example the component style should override the page styles and the background should be yellow. ie. the dev mode is correct. Worryingly it's the production mode that gets it wrong.

It works for me, thanks)

**Comment by simplenotezy** [comment]
> I'm not even sure what to do as a workaround for this.
> 
> I think an easy solution would be to automatically apply page-specific identifiers as part of the css modules hash generation, but the ability to configure that isn't exposed.
> 
> In my app today, I guess I could pass the `style` attribute and inline my styles, but I have a heavy reliance on PostCSS stuff that can't be leveraged in my JS.
> 
> I tried using `classNames` with dynamic values and `.bind()` to no avail... Even `!important` wont work because if more pages are passing one-offs to the same components, those will be fighting eachother.

A workaround could be more specific selectors, e.g.:

Instead of: `.Button {}`, do `.Container .Button` - that should fix it.

**Comment by rachelslurs** [comment]
> > I'm not even sure what to do as a workaround for this.
> > I think an easy solution would be to automatically apply page-specific identifiers as part of the css modules hash generation, but the ability to configure that isn't exposed.
> > In my app today, I guess I could pass the `style` attribute and inline my styles, but I have a heavy reliance on PostCSS stuff that can't be leveraged in my JS.
> > I tried using `classNames` with dynamic values and `.bind()` to no avail... Even `!important` wont work because if more pages are passing one-offs to the same components, those will be fighting eachother.
> 
> A workaround could be more specific selectors, e.g.:
> 
> Instead of: `.Button {}`, do `.Container .Button` - that should fix it.

I actually went ahead and tried doing this using a postcss plugin for a fairly large codebase, and didn't experience much luck given that the postcss seems to not be the last loader to modify the css classnames. See my bug here if you're curious: https://github.com/vercel/next.js/issues/25951

**Comment by ErAz7** [comment]
A good work around to avoid style order issues could be using SCSS modules and nest styles of each component in a `.root` selector like this:

**home.tsx**
```jsx
export default function HomePage() {
   return (
      <div className={styles.root}>
          <SomeSharedComponent className={styles.shared} />
      </div>
   )
}
```

**home.modules.scss**
```scss
.root {
   .shared {
      color: green;
   }
}
```

This gives the green color more specificity, thus overrides any styles coming from `SomeSharedComponent` resulted by loading css in wrong order

**Comment by kylemh** [comment]
@timneutkens / @Timer im guessing this is a "won't fix" until Turbopack?

**Comment by belgattitude** [comment]
Would love to know too. Thx

**Comment by coskuncakir** [comment]
having the same problem.. is there any way to fix this?

**Comment by Enkratia** [comment]
This actualy may be caused by Next.js because `CSS modules` are working perfectly until i use "use client" in ANY component. 

When i use `use client` in one component  =>  order of `CSS modules` stop to work in entire project. 
When i use `use client` in this component  =>  order of `CSS modules`  works again in entire project. 

P.S.: I have met this problem with Vite some time ago, i don`t know what the cause of problem was there.

**Comment by waldell** [comment]
Would love to hear if this is gonna be fixed? I spent hours on debug and researching until I found the issue, googled it and found this issue. 
I mean, you really should expect this kind of basic thing to work in such a large library as Next

**Comment by jackvividseats** [comment]
This has been causing an issue for me as well. I have no productive solution just wanted to bump the discussion.

**Comment by fresonn** [comment]
Really, it's still relevant?

**Comment by Enkratia** [comment]
If someone is still interested on the topic, those can check last paragraph of my last message (about aliases).

**Comment by yunsii** [comment]
https://github.com/vercel/next.js/issues/10148
https://github.com/vercel/next.js/issues/16630

And this issue all are css order related bugs, but it still not work as expected fot now 😦

**Comment by theenoahmason** [comment]
I just spent hours upon hours messing with this, and to be honest it has driven me fully bananas 🍌. 

Its the first time i've experiences this out of several Next Apps, and it is the first time I am using the app router on a real project, and this issue raised it's head pretty early on in the process. Below are my findings and how I fixed it for now. 

**What causes it**

Any time components are imported into the `layout` component, we get both the unordered css classes, and duplication of classes. As I was experimenting, it became pretty clear that if no components are imported into the `layout` component, **BOTH** issues go away, at least in my minimal example. I tried messing with `@layers` as well, but even those are unordered in an unpredictable way, EVEN if the layer order is specified at the top of the global style file imported in the `layout` component.

**How I "Fixed" it**

I stripped most of everything out of the `layout` component, and instead created a `Page` component, that serves as the global layout. This is where my layout components like `Header`, `Nav`, `Footer` components go. This component is then used as the base component wrapping all next pages. The caveat is that the `Page` component is now re-rendered on route changes, but I am experiencing no negative effects other than trivial things like a button hover animation not finishing before the new page renders. **The main thing is that this issue is no longer present - to be honest its not really possible to build an app with custom styles without doing this IMO**. See code below.

**Notes on the code below:** I am not using modules (cause I don't like em), and I kept the layers because it still created nice segmentation when debugging, but I don't even think its necessary.

**app/layout**
```tsx
import type { Metadata } from 'next';
import classNames from 'classnames';
import Navigation from '@/components/Navigation/Navigation';
import Header from '@/components/Header/Header';
import { Inter_Tight } from 'next/font/google';
import './global.scss';

const interTight = Inter_Tight({ subsets: ['latin'] });

export interface LayoutProps {
  children?: React.ReactNode;
}

export default async function Layout(props: LayoutProps) {
  const { children } = props;
  const rootClass = classNames('Layout', interTight.className);
  return (
    <html lang="en">
      <body className={rootClass}>
        {children}
      </body>
    </html>
  );
}

export const metadata: Metadata = {
  title: 'Next.js App',
  description: 'An example Next.js App.',
}

```

**components/Page**
```tsx
import classNames from 'classnames';
import PageHeader from '@/components/PageHeader/PageHeader';
import Footer from '@/components/Footer/Footer';
import Navigation from '@/components/Navigation/Navigation';
import { Auth } from '@/utils/Auth/Auth';
import Header from '@/components/Header/Header';
import './Page.scss';

export interface PageProps {
    id?: string;
    className?: string;
    children?: React.ReactNode;
    title: string;
}

export default async function Page(props: PageProps) {
    const { id, className, children, title } = props;
    const rootClass = classNames('Page', className);
    const session = await Auth.getServerSession();
    return (
        <div id={id} className={rootClass}>
            <Header title="Survival Fantasy" />
            <main className="Page__main">
                <aside className="Page__sidebar">
                    {session?.user ? (
                        <Navigation lists={[
                            { 
                                label: 'Navigation',
                                items: [
                                    { children: 'Sign Out', href: '/sign-out' },
                                ],
                            }
                        ]} />
                    ) : (
                        <Navigation lists={[
                            { 
                                label: 'Navigation',
                                items: [
                                    { children: 'Sign Up', href: '/sign-up' },
                                    { children: 'Sign In', href: '/sign-in' },
                                ],
                            }
                        ]} />
                    )}
                </aside>
                <div className="Page__content">
                    <div className="Page__window">
                        <PageHeader title={title} />
                        <div className="Page__outlet">
                            {children}
                        </div>
                    </div>
                    <Footer />
                </div>
            </main>
        </div>
    );
}
```

**app/page.tsx**
```tsx
import { Metadata } from 'next';
import classNames from 'classnames';
import Page from '@/components/Page/Page';

export default async function HomePage() {
  const rootClass = classNames('HomePage');
  return (
    <Page title="Home" className={rootClass}>
      Hello, World!
    </Page>
  );
}

export const metadata: Metadata = {
  title: 'Home',
  description: 'Hello, world!',
}
```

**app/global.scss**
```scss
@layer root, normalize, reset, primitives, components;

@import '/sass/root.scss';

@layer normalize {
    @import 'normalize.css/normalize';
}

@import '/sass/reset';
```

**Comment by harryfinn** [comment]
Have been encountering the same issue outline both here and previously in https://github.com/vercel/next.js/issues/16630.

Using app router on Next.js v14.0.4 I am unable to override css vars from an imported external package, even when ordering my overrides below the imported package.

The only work around I've found to work both in local dev and generated prod build is the one outline here https://github.com/vercel/next.js/issues/16630#issuecomment-1762463616

Very keen to know if there are any alternatives at this stage, or whether it's a waiting game for Turbopack at this point?

**Comment by landsman** [comment]
This is real pain.

**Comment by thexpand** [comment]
Happens again in v14.2.2

**Comment by r1tsuu** [comment]
can confirm the same on 14.3.0-canary-7.. this is so annoying when your feature works perfectly locally but on production not and then people are pinging at you

**Comment by samcx** [comment]
Closing this since this fix was landed → https://github.com/vercel/next.js/issues/51030#issuecomment-2005254907.

**Comment by thexpand** [comment]
@samcx I don't understand why you're closing this when it hasn't been fixed. Literally, the issue you shared has comments of people who tried the latest version and it still breaks.

**Comment by samcx** [comment]
@thexpand There are different CSS-related issues shared in that thread, so I haven't closed that one. The PR in that comment specifically describes a fix for CSS Modules ordering.

The [version you shared](https://github.com/vercel/next.js/releases/tag/v14.2.2) I believe does not have that PR fix, but it will be in the latest canary.

**Comment by thexpand** [comment]
Oh, okay, but the comment you shared is misleading - it says the fix landed on `v14.2.0-canary.28`, which I thought was released with 14.2.0, but maybe it wasn't.
